lib/ansible/module_utils/fortios.py

```diff
@@ -198,3 +208,819 @@ def get_error_infos(self, cli_errors):
+
+    def _update_temporary_and_permanent_objects(self, failures):
+        update_objects = [o for o in self._update_config if o[
+            self._object_identifier] in self._object_ids_to_update]
+        for forti_object in update_objects:
+            if not self._check_mode:
+                response = self._edit(
+                    self._endpoint + "/%s" % str(forti_object[self._object_identifier]), data=forti_object)
+                if response['http_status'] != 200:
+                    failures[forti_object[self._object_identifier]
+                             ] = self.http_status_codes[response['http_status']]
+
+    def _reset_permanent_objects(self, failures):
+        for identifier in self._permanent_object_ids_to_reset:
+            response_data = deepcopy(self._default_object_configuration)
+            response_data[self._object_identifier] = identifier
+            if not self._check_mode:
+                response = self._edit(self._endpoint + "/%s" %
+                                      str(identifier), data=response_data)
+                if response['http_status'] != 200:
+                    failures[identifier] = self.http_status_codes[
+                        response['http_status']]
+
+    def _get_current_object(self, forti_object):
+        for obj in self._fortigate_current_config:
+            if obj[self._object_identifier] == forti_object[self._object_identifier]:
+                return obj
+        return None
+
+    def _create_new_objects(self):
+        if not self._update_config:
+            return
+        existing_object_ids = [o[self._object_identifier]
+                               for o in self._fortigate_current_config]
+        new_objects = [o for o in self._update_config if o[self._object_identifier] not in existing_object_ids and
+                       o[self._object_identifier] not in self._permanent_object_ids]
+        failures = {}
+        for forti_object in new_objects:
+            if not self._check_mode:
+                response = self._create(self._endpoint, data=forti_object)
+                if response['http_status'] != 200:
+                    failures[forti_object[self._object_identifier]
+                             ] = self.http_status_codes[response['http_status']]
+        if failures:
+            self.fail("Failed to create objects:\n ", msg_args=failures)
+
+    def _update_single_object_endpoint(self):
+
+        if self._update_config and isinstance(self._update_config, list):
+            self._update_config = self._update_config[0]
+        else:
+            self._update_config = self._default_object_configuration
+
+        if not self._check_mode:
+            self._edit(self._endpoint, data=self._update_config)
+
+    def __enter__(self):
+        return self
+
+    def __del__(self):
+        try:
+            self._logout()
+        except AttributeError:
+            pass
+
+    def __exit__(self, *args):
+        pass
```

Comment:
You seem to have gone through the process of making this class a context manager, but don't use it as a context manager.

Reply:
The only thing I want to do is to logout when the execution is complete or failed. I believe both __del__ and __exit__ are being called and to avoid having duplicate logout calls I just ended up doing it in __del__ (line 507)